cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/assumptions/Assumption.kt

```diff
@@ -0,0 +1,261 @@
+    var assumptionID: String? = null,
+    @Relationship(value = "ASSUMPTION_SCOPE", direction = Relationship.Direction.OUTGOING)
+    var assumptionScope: Node? = node,
+    var status: AssumptionStatus = AssumptionStatus.Ignored,
+) : OverlayNode() {
+
+    init {
+        super.underlyingNode = node ?: edge?.start
+        if (super.underlyingNode == null && assumptionID == null) {
+            log.warn(
+                "Creating an assumption with no associated node or edge requires setting and assumptionID but was null"
+            )
+        }
+
+        if (assumptionScope == null) {
+            this@Assumption.assumptionScope = super.underlyingNode
+        }
+        name = Name(assumptionType.name)
+        location = node?.location
+    }
+}
+
+/**
+ * the [AssumptionStatus] that an assumption can have, default being set to
+ * [AssumptionStatus.Undecided]. When an assumption is [AssumptionStatus.Accepted] or
+ * [AssumptionStatus.Ignored], the object it is attached to is considered correct, e.g. nodes,
+ * edges, paths are considered to exist and results are true or false, if based on a contradiction
+ * result. If it is [AssumptionStatus.Rejected], nodes and edges containing them are supposed to be
+ * considered nonexistent, or not existing with the same properties. Results with assumptions and
+ * status [AssumptionStatus.Rejected] should be considered false positives.
+ */
+enum class AssumptionStatus {
+    /**
+     * Default state: Patterns(Paths) depending on the assumption are returned, result has an
+     * undecided state, assumptions are shown.
+     */
+    Undecided,
+    /**
+     * User or Algorithm: Patterns(Paths) depending on the assumption are not returned, results have
+     * a decided state, assumptions are (not?) shown.
+     */
+    Rejected,
+    /**
+     * User or Algorithm: Patterns(Paths) depending on the assumption are returned, results have a
+     * decided state, assumptions are shown.
+     */
+    Accepted,
+    /**
+     * User or Algorithm: Patterns(Paths) depending on the assumption are returned, results have a
+     * decided state, assumptions are not shown.
+     */
+    Ignored,
+}
+
+enum class AssumptionType {
+    SyntaxAmbiguityAssumption,
+    InferenceAssumption,
+    ClosedMacroAssumption,
+    UnsupportedLanguageProblem,
+    MissingCodeProblem,
+    AmbiguityAssumption,
+    ConceptAssumption,
+    ExhaustiveEnumerationAssumption,
+    CompletenessAssumption,
+    SoundnessAssumption,
+    ControlFlowAssumption,
+    CFIntegrityAssumption,
+    NoExceptionsAssumption,
+    CFAllOrNothingExecutesAssumption,
+    TrustedConfigAssumption,
+    DataFlowAssumption,
+    ExternalDataAssumption,
+    NetworkAvailableAssumption,
+    ResourceExistsAssumption,
+    ServiceReachableAssumption,
+    AtomicExecutionAssumption,
+    TrustBoundaryAssumption,
+    DataRangeAssumption,
+    TrustedInputAssumption,
+}
+
+interface HasAssumptions {
+    val assumptionNodes: MutableList<Assumption>
```

Comment:
as discussed offline, rename to `assumptions`, we will rename/remove/convert the assumptions in the security goal

Reply:
Rename the `assumptions` of Security Goal to `legacyAssumptions`. We will then later adjust the YAML files.